Enable String Deduplication on JDK 17+

schlosna opened the pull request:
## Before this PR
JVM string deduplication was not enabled

## After this PR
Reduce heap size by allowing JVM garbage collectors to deduplicate strings.
==COMMIT_MSG==
Enable String Deduplication on JDK 17+

G1 and Shenandoah GC support string deduplication as of JDK 17
* https://openjdk.java.net/jeps/192
* https://bugs.openjdk.org/browse/JDK-8264718

Only enable on JDK 17+ due to fix
https://bugs.openjdk.org/browse/JDK-8277981

JDK 18+ enable string deduplication for SerialGC, ParallalGC, and ZGC.
https://malloc.se/blog/zgc-jdk18
==COMMIT_MSG==

Closes https://github.com/palantir/sls-packaging/issues/1378

## Possible downsides?
Enabling this by default may trigger some unknown JDK bug(s).

carterkozak commented:
Let's pick a couple performance-sensitive canary services and try this out before including in sls-packaging

iamdanfox commented:
we've run with this one one apollo node for 24 hours and interestingly haven't seen a noticeable reduction in heap yet (but also haven't seen any _downside_ in terms of CPU either).  Perhaps we should trial it on a _really_ performance sensitive service like MP or timelock before blasting to everyone?

carterkozak commented:
@pkoenig10 Thoughts on giving this a shot in MP?

carterkozak commented:
Happy to merge this. Currently this branch targets #1374 which is still do-not-merge, up to you how you'd like to sequence things.

schlosna commented:
I will rebase off develop later tonight/tomorrow and update this to target develop

schlosna committed:
Enable String Deduplication on JDK 17+

G1 and Shenandoah GC support string deduplication as of JDK 17
  * https://openjdk.java.net/jeps/192
  * https://bugs.openjdk.org/browse/JDK-8264718

Only enable on JDK 17+ due to fix https://bugs.openjdk.org/browse/JDK-8277981

JDK 18+ enable string deduplication for SerialGC, ParallalGC, and ZGC.
See https://malloc.se/blog/zgc-jdk18

svc-changelog committed:
Add generated changelog entries

schlosna committed:
spacing

carterkozak reviewed (approved):
Thanks!

svc-autorelease commented:
Released [7.32.0](https://github.com/palantir/sls-packaging/releases/tag/7.32.0)
